File: api/Services/TrainRegistry.cs
using System.Collections.Concurrent;
using Api.Models;

namespace Api.Services;

public class TrainRegistry
{
    private record TrainEntry(
        string Headcode,
        string? Toc,
        string? Destination,
        int DelayMinutes,
        TdBerthPosition? Berth,
        double? Lat,
        double? Lng,
        DateTime UpdatedAt
    );

    private readonly ConcurrentDictionary<string, TrainEntry> _trains = new();
    private readonly TrainStateService _state;

    private int _tdMessages;
    private int _darwinMessages;
    private int _berthHits;
    private int _berthMisses;
    private string? _lastTdSample;
    private string? _lastDarwinSample;

    public TrainRegistry(TrainStateService state)
    {
        _state = state;
    }

    public object GetDebugInfo() => new
    {
        TdMessages = _tdMessages,
        DarwinMessages = _darwinMessages,
        BerthHits = _berthHits,
        BerthMisses = _berthMisses,
        TrainsWithPosition = _trains.Values.Count(t => t.Lat.HasValue),
        TrainsTotal = _trains.Count,
        LastTdSample = _lastTdSample,
        LastDarwinSample = _lastDarwinSample,
    };

    public void UpdateFromTd(string headcode, TdBerthPosition berth, double? lat, double? lng)
    {
        Interlocked.Increment(ref _tdMessages);
        if (lat.HasValue) Interlocked.Increment(ref _berthHits);
        else Interlocked.Increment(ref _berthMisses);

        if (_tdMessages <= 5)
            _lastTdSample = $"headcode={headcode} area={berth.AreaId} berth={berth.BerthId} lat={lat} lng={lng}";

        var entry = _trains.AddOrUpdate(
            headcode,
            _ => new TrainEntry(headcode, null, null, 0, berth, lat, lng, DateTime.UtcNow),
            (_, existing) => existing with { Berth = berth, Lat = lat, Lng = lng, UpdatedAt = DateTime.UtcNow }
        );

        if (lat.HasValue && lng.HasValue)
            PublishPosition(headcode, entry);
    }

    public void UpdateFromDarwin(string headcode, string? toc, string? destination, int delayMinutes)
    {
        Interlocked.Increment(ref _darwinMessages);
        if (_darwinMessages <= 3)
            _lastDarwinSample = $"headcode={headcode} toc={toc} dest={destination} delay={delayMinutes}";

        var entry = _trains.AddOrUpdate(
            headcode,
            _ => new TrainEntry(headcode, toc, destination, delayMinutes, null, null, null, DateTime.UtcNow),
            (_, existing) => existing with { Toc = toc, Destination = destination, DelayMinutes = delayMinutes }
        );

        if (entry.Lat.HasValue && entry.Lng.HasValue)
            PublishPosition(headcode, entry);
    }

    public void RemoveFromTd(string headcode)
    {
        if (_trains.TryRemove(headcode, out _))
            _state.RemoveTrain(headcode);
    }

    private void PublishPosition(string headcode, TrainEntry entry)
    {
        if (!entry.Lat.HasValue || !entry.Lng.HasValue) return;

        var position = new TrainPosition(
            headcode,
            headcode,
            entry.Toc,
            entry.Lat.Value,
            entry.Lng.Value,
            entry.Berth?.AreaId,
            entry.Berth?.BerthId,
            entry.Destination,
            entry.DelayMinutes,
            entry.UpdatedAt
        );

        _state.UpdateTrain(position);
    }
}

File: frontend/lib/services/train-registry.ts
import type { TrainStateService } from './train-state-service'
import type { TrainPosition } from '../../types/train'

interface TdBerth {
  areaId: string
  berthId: string
}

interface TrainEntry {
  rid: string
  headcode: string | null
  toc: string | null
  origin: string | null
  destination: string | null
  delayMinutes: number
  berth: TdBerth | null
  lat: number | null
  lng: number | null
  destLat: number | null
  destLng: number | null
  updatedAt: string
}

export interface DebugInfo {
  tdMessages: number
  darwinMessages: number
  berthHits: number
  berthMisses: number
  trainsWithPosition: number
  trainsTotal: number
  lastTdSample: string | null
  lastDarwinSample: string | null
}

export class TrainRegistry {
  private readonly trains = new Map<string, TrainEntry>()
  private readonly headcodeToRid = new Map<string, string>()
  private readonly state: TrainStateService

  private tdMessages = 0
  private darwinMessages = 0
  private berthHits = 0
  private berthMisses = 0
  private lastTdSample: string | null = null
  private lastDarwinSample: string | null = null

  constructor(state: TrainStateService) {
    this.state = state
  }

  getDebugInfo(): DebugInfo {
    return {
      tdMessages: this.tdMessages,
      darwinMessages: this.darwinMessages,
      berthHits: this.berthHits,
      berthMisses: this.berthMisses,
      trainsWithPosition: [...this.trains.values()].filter(t => t.lat !== null).length,
      trainsTotal: this.trains.size,
      lastTdSample: this.lastTdSample,
      lastDarwinSample: this.lastDarwinSample,
    }
  }

  updateFromTd(headcode: string, berth: TdBerth, lat: number | null, lng: number | null) {
    this.tdMessages++
    if (lat !== null) this.berthHits++
    else this.berthMisses++

    if (this.tdMessages <= 5)
      this.lastTdSample = `headcode=${headcode} area=${berth.areaId} berth=${berth.berthId} lat=${lat} lng=${lng}`

    const rid = this.headcodeToRid.get(headcode) ?? headcode
    const existing = this.trains.get(rid)

    const entry: TrainEntry = existing
      ? { ...existing, headcode, berth, lat, lng, updatedAt: new Date().toISOString() }
      : { rid, headcode, toc: null, origin: null, destination: null, delayMinutes: 0, berth, lat, lng, destLat: null, destLng: null, updatedAt: new Date().toISOString() }

    this.trains.set(rid, entry)

    if (lat !== null && lng !== null)
      this.publishPosition(rid, entry)
  }

  updateFromDarwin(
    rid: string,
    headcode: string | null,
    toc: string | null,
    origin: string | null,
    destination: string | null,
    delayMinutes: number,
    destLat?: number | null,
    destLng?: number | null,
  ) {
    this.darwinMessages++
    if (this.darwinMessages <= 5)
      this.lastDarwinSample = `rid=${rid} headcode=${headcode} toc=${toc} origin=${origin} dest=${destination} delay=${delayMinutes}`

    if (headcode) this.headcodeToRid.set(headcode, rid)

    const existing = this.trains.get(rid)
    const entry: TrainEntry = existing
      ? {
          ...existing,
          headcode: headcode ?? existing.headcode,
          toc: toc ?? existing.toc,
          origin: origin ?? existing.origin,
          destination: destination ?? existing.destination,
          delayMinutes,
          destLat: destLat ?? existing.destLat,
          destLng: destLng ?? existing.destLng,
        }
      : { rid, headcode, toc, origin, destination, delayMinutes, berth: null, lat: null, lng: null, destLat: destLat ?? null, destLng: destLng ?? null, updatedAt: new Date().toISOString() }

    this.trains.set(rid, entry)

    if (entry.lat !== null && entry.lng !== null)
      this.publishPosition(rid, entry)
  }

  removeFromTd(headcode: string) {
    const rid = this.headcodeToRid.get(headcode) ?? headcode
    if (this.trains.delete(rid))
      this.state.removeTrain(rid)
  }

  removeByRid(rid: string) {
    if (this.trains.delete(rid))
      this.state.removeTrain(rid)
  }

  private publishPosition(rid: string, entry: TrainEntry) {
    if (entry.lat === null || entry.lng === null) return

    let heading: number | null = null
    if (entry.destLat !== null && entry.destLng !== null)
      heading = bearingDeg(entry.lat, entry.lng, entry.destLat, entry.destLng)

    const position: TrainPosition = {
      rid,
      headcode: entry.headcode ?? rid,
      toc: entry.toc,
      lat: entry.lat,
      lng: entry.lng,
      fromTiploc: entry.berth?.areaId ?? null,
      toTiploc: entry.berth?.berthId ?? null,
      origin: entry.origin,
      destination: entry.destination,
      delayMinutes: entry.delayMinutes,
      heading,
      updatedAt: entry.updatedAt,
    }

    this.state.updateTrain(position)
  }
}

function bearingDeg(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const dLng = (lng2 - lng1) * Math.PI / 180
  const φ1 = lat1 * Math.PI / 180
  const φ2 = lat2 * Math.PI / 180
  const y = Math.sin(dLng) * Math.cos(φ2)
  const x = Math.cos(φ1) * Math.sin(φ2) - Math.sin(φ1) * Math.cos(φ2) * Math.cos(dLng)
  return (Math.atan2(y, x) * 180 / Math.PI + 360) % 360
}
